lib/Mojo/mysql/Results.pm

```diff
@@ -42,34 +42,53 @@ sub state { shift->sth->state }
 sub _c { Mojo::Collection->new(@_) }
 
 sub _expand {
-  my ($self, @rows) = @_;
-
-  return @rows unless my $mode = $self->{expand} and $rows[0];
-
-  # Force expanding
-  return map {
-    my $r = $_;
-    $_ = from_json $_ for grep {/^(\[|\{).*(\}|\])/} values %$r;
-    $r;
-  } @rows if $mode == 2;
-
-  # Only expand json columns
-  my ($idx, $name) = @$self{qw(idx name)};
-  unless ($idx) {
-    my $types = $self->db->mysql->_dbi_attr($self->sth, 'type');
-    my @idx   = grep { $types->[$_] == 245 or $types->[$_] == 252 } 0 .. $#$types;    # 245 = MySQL, 252 = MariaDB
-    ($idx, $name) = @$self{qw(idx name)} = (\@idx, [@{$self->columns}[@idx]]);
-  }
+  my ($self, $to) = @_;
+
+  # Get field names and types, needs to be done before reading from sth
+  my $mode = $self->{expand} || 0;
+  my ($idx, $names) = $mode == 1 ? $self->_types : ();
 
-  return @rows unless @$idx;
-  if (ref $rows[0] eq 'HASH') {
-    for my $r (@rows) { $r->{$_} and ($r->{$_} = from_json $r->{$_}) for @$name }
+  # Fetch sql data
+  my $sql_data;
+  if ($to->{list}) {
+    $sql_data = $to->{hash} ? $self->sth->fetchall_arrayref({}) : $self->sth->fetchall_arrayref;
   }
   else {
-    for my $r (@rows) { $r->[$_] and ($r->[$_] = from_json $r->[$_]) for @$idx }
+    @$sql_data = $to->{hash} ? $self->sth->fetchrow_hashref : $self->sth->fetchrow_arrayref;
+  }
+
+  # expand(), only expand json columns
+  if ($mode == 1) {
```

Comment:
Based on your comments on which part of the code that actually takes time to run, I suggest this:

```
sub _expand {
  ...
  my $from_json = __PACKAGE__->can(sprintf '_from_json_mode_%s_%s', $mode, $to->{hash} ? 'hash' : 'array');
  $from_json->($_, $idx, $names) for @$sql_data;
  ...
}

sub _from_json_mode_1_array {
  my ($r, $idx, $names) = @_;
  $r->[$_] = from_json $r->[$_] for grep { defined $r->[$_] } @$idx;
}

sub _from_json_mode_1_hash {
  my ($r, $idx, $names) = @_;
  $r->{$_} = from_json $r->{$_} for grep { defined $r->{$_} } @$names;
}

sub _from_json_mode_2_array {
  my ($r, $idx, $names) = @_;
  $_ = from_json $_ for grep /^[\[{].*[}\]]$/, @$r;
}

sub _from_json_mode_2_hash {
  my ($r, $idx, $names) = @_;
  $_ = from_json $_ for grep /^[\[{].*[}\]]$/, values %$r;
}
```